tidy3d/plugins/smatrix/smatrix.py

```diff
@@ -48,7 +49,31 @@ class Port(Box):
 
 MatrixIndex = Tuple[str, pd.NonNegativeInt]  # the 'i' in S_ij
 Element = Tuple[MatrixIndex, MatrixIndex]  # the 'ij' in S_ij
-SMatrixType = Dict[MatrixIndex, Dict[MatrixIndex, complex]]  # the 'S' itself
+
+
+class SMatrixDataArray(DataArray):
+    """Scattering matrix elements.
+
+    Example
+    -------
+    >>> port_in = ['port1', 'port2']
+    >>> port_out = ['port1', 'port2']
+    >>> mode_index_in = [0, 1]
+    >>> mode_index_out = [0, 1]
+    >>> f = [2e14]
+    >>> coords = dict(
+    ...     port_in=ports_in,
+    ...     port_out=ports_out,
+    ...     mode_index_in=mode_index_in,
+    ...     mode_index_out=mode_index_out,
+    ...     f=f
+    ... )
+    >>> fd = SMatrixDataArray((1 + 1j) * np.random.random((2, 2, 2, 2, 1)), coords=coords)
+    """
+
+    __slots__ = ()
+    _dims = ("port_in", "port_out", "mode_index_in", "mode_index_out", "f")
```

Comment:
I'd change the ordering here to `("port_out", "mode_index_out", "port_in", "mode_index_in", "f")` to be more compatible with a matrix representation if we use `SMatrixDataArray.values`. The usual ordering of indices is output first, input second.